### PR TITLE
Write script to generate report for routes

### PR DESCRIPTION
**Pull request by kevinchan159**
We currently support two kinds of report generations.
- `routes_report.csv`: csv of `timestamp, origin, destination`
- `search_count_report.csv`: csv of `day, hour, count`

### CODE REVIEW

File: generate_report.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+from src.search_count_report import SearchCountReport
+
+def gen_report(filename):
+  # Make sure filename is a csv file
+  extension_idx = filename.rfind('.csv')
+  if extension_idx == -1:
+    print(f'{filename} is not a csv file! Please try again')
+    return
+
+  # Read filename as csv
```
Comment (alanna-zhou) [inline generate_report.py]: should we move this to a separate method called `get_csv_reader`

File: generate_report.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  csv_file = open(filename)
+  csv_reader = csv.reader(csv_file, delimiter=',')
+  
+  # Setup writing to routes_report.csv
```
Comment (alanna-zhou) [inline generate_report.py]: and this to a `get_csv_writer`

File: generate_report.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import csv
+import sys
+
+from src.routes_report import RoutesReport
+from src.search_count_report import SearchCountReport
+
+def gen_report(filename):
```
Comment (alanna-zhou) [inline generate_report.py]: should this be `file_name` for consistency since you use `report_name` instead of `reportname` as a variable below

File: src/routes_report.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  def generate(self):
+    # Write fields to routes_report.csv 
+    self.writer.writerow(['timestamp', 'origin', 'destination'])
+
+    # Read routes.csv 
+    for index, row in enumerate(self.reader):
+      if index == 0:
+        continue
+      msg_json = utils.parse_msg_json(row[1])
+      origin, dest = self.get_origin_dest(msg_json)
```
Comment (alanna-zhou) [inline src/routes_report.py]: shouldn't this also be in `src.utils` since we might need to use `get_origin_dest` in other types of reports?

File: src/search_count_report.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+from src.report import Report
+import src.utils as utils
+
+class SearchCountReport(Report):
+
+  def generate(self):
+    # Write fields to search_count.csv 
+    self.writer.writerow(['day', 'hour', 'count'])
+
+    time_to_count_dict = {}
```
Comment (alanna-zhou) [inline src/search_count_report.py]: should we rename this to `day_hour_to_count_dict`

File: src/search_count_report.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        else:
+          time_to_count_dict[day][hour] = 1
+      else:
+        time_to_count_dict[day] = {hour: 1}
+
+    for day in time_to_count_dict:
+      for hour in time_to_count_dict[day]:
+        self.writer.writerow([day, hour, time_to_count_dict[day][hour]])
+
+  def parse_date(self, timestamp):
```
Comment (alanna-zhou) [inline src/search_count_report.py]: should we also move this to `src.utils` or do you think it's not necessary

File: src/search_count_report.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    suffixes = ['st', 'th', 'rd']
+    time_str = timestamp
+    # Strip off ordinal suffixes so we can easily convert to datetime object
+    for s in suffixes:
+      time_str = time_str.replace(s, '')
+    # Strip the milliseconds off of the time
+    time_str = time_str.split('.')[0]
+    return datetime.strptime(time_str, "%B %d %Y, %H:%M:%S")
+
+  def parse_day(self, timestamp):
```
Comment (alanna-zhou) [inline src/search_count_report.py]: `parse_day_str_from_timestamp`? or `parse_day_str_from_str`? or too wordy lmk

File: src/search_count_report.py (hunk 6 above)
Comment (alanna-zhou) [inline src/search_count_report.py]: `parse_date_from_timestamp`? or `parse_date_from_str`?

File: src/utils.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+def parse_msg_json(msg):
+  """This parses the json string from the message field.
+  """
```
Comment (alanna-zhou) [inline src/utils.py]: why are the docs ending quotes on the next line. is this python standard? it looks kinda wacky

File: src/utils.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    return ''
+  return msg[start_idx:]
+
+def parse_key(msg_json, key):
+  """This parses the value corresponding to [key] from the msg_json.
+  """
+  key_idx = msg_json.find(key)
+  if key_idx == -1:
+    return ''
+  key_start_idx = key_idx + len(key) + 3
```
Comment (alanna-zhou) [inline src/utils.py]: why are you doing this

Comment (alanna-zhou) [inline src/utils.py]: should you make `3` a constant somewhere or is this too trivial to do that bc i have no idea why you need to add 3

**Review by alanna-zhou — APPROVED** [review]
hopefully we can merge this in asap before it's an open pr for a month

File: generate_report.py (hunk 3 above)
Comment (young-k) [reply generate_report.py]: isn't `filename` a word though?

Comment (young-k) [reply generate_report.py]: also prefer `generate_report`

File: generate_report.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  # Make sure filename is a csv file
+  extension_idx = filename.rfind('.csv')
```
Comment (young-k) [inline generate_report.py]: `idx`?

File: src/routes_report.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+      timestamp = row[0]
+      self.writer.writerow([timestamp, origin, dest])
+
+  def get_origin_dest(self, msg_json):
+    origin = utils.parse_key(msg_json, 'originName')
+    dest = utils.parse_key(msg_json, 'destinationName')
+    if origin == 'Current Location' or origin == '':
```
Comment (young-k) [inline src/routes_report.py]: `if not origin or origin == 'Current Location'` is more python-y

File: src/utils.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+  """
+  start_idx = msg.find('{')
+  if start_idx == 1:
+    return ''
+  return msg[start_idx:]
+
+def parse_key(msg_json, key):
+  """This parses the value corresponding to [key] from the msg_json.
+  """
+  key_idx = msg_json.find(key)
```
Comment (young-k) [inline src/utils.py]: `idx`?

File: src/utils.py [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+  """This parses the json string from the message field.
+  """
+  start_idx = msg.find('{')
+  if start_idx == 1:
+    return ''
+  return msg[start_idx:]
+
+def parse_key(msg_json, key):
+  """This parses the value corresponding to [key] from the msg_json.
+  """
```
Comment (young-k) [inline src/utils.py]: this `"""` should be on the previous line

**Review by young-k — CHANGES_REQUESTED** [review]
Python code should be indented by 4, address @alanna-zhou's changes as well

File: generate_report.py (hunk 1 above)
Comment (kevinchan159) [reply generate_report.py]: i don't think its necessary since we only do it once and its 2 lines

File: src/routes_report.py (hunk 4 above)
Comment (kevinchan159) [reply src/routes_report.py]: will move it once that comes up

File: src/search_count_report.py (hunk 5 above)
Comment (kevinchan159) [reply src/search_count_report.py]: i think `time_to_count_dict` is clearer

File: src/search_count_report.py (hunk 6 above)
Comment (kevinchan159) [reply src/search_count_report.py]: we can move it if we need it for other reports

File: src/search_count_report.py (hunk 7 above)
Comment (kevinchan159) [reply src/search_count_report.py]: i think parse_day is clear enough

File: generate_report.py (hunk 10 above)
Comment (kevinchan159) [reply generate_report.py]: `index`

File: generate_report.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
+import csv
+import sys
+
+from src.routes_report import RoutesReport
+from src.search_count_report import SearchCountReport
+
+def generate_report(filename):
+    # Make sure filename is a csv file
+    extension_idx = filename.rfind('.csv')
```
Comment (young-k) [inline generate_report.py]: prefer `index`

File: generate_report.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    if extension_idx == -1:
+        print(f'{filename} is not a csv file! Please try again.')
+        return
+
+    # Setup reader
```
Comment (young-k) [inline generate_report.py]: i think this comment is redundant

File: generate_report.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+    csv_file = open(filename)
+    csv_reader = csv.reader(csv_file, delimiter=',')
```
Comment (young-k) [inline generate_report.py]: the default delimiter is `,` according to https://docs.python.org/3.6/library/csv.html#csv-fmt-params, can we remove this everywhere since it's redundant?

File: src/search_count_report.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+            datetime = self.parse_date(timestamp)
+            hour = datetime.hour
+            day = self.parse_day(timestamp)
+            if day in time_to_count_dict:
+                if hour in time_to_count_dict[day]:
+                    time_to_count_dict[day][hour] += 1
+                else:
+                    time_to_count_dict[day][hour] = 1
+            else:
+                time_to_count_dict[day] = {hour: 1}
```
Comment (young-k) [inline src/search_count_report.py]: i feel like there's a better way to write this, this `if else` pair is pretty clunky to read

File: src/search_count_report.py [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+        for day in time_to_count_dict:
+            for hour in time_to_count_dict[day]:
+                self.writer.writerow([day, hour, time_to_count_dict[day][hour]])
+
+    def parse_date(self, timestamp):
+        suffixes = ['st', 'th', 'rd']
```
Comment (young-k) [inline src/search_count_report.py]: comment explaining these suffixes?  call it `ordinal_suffixes` if that's what they are

File: src/search_count_report.py [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+        time_str = timestamp
+        # Strip off ordinal suffixes so we can easily convert to datetime object
+        for s in suffixes:
+            time_str = time_str.replace(s, '')
+        # Strip the milliseconds off of the time
+        time_str = time_str.split('.')[0]
```
Comment (young-k) [inline src/search_count_report.py:44]: if the above two operations are actually `strip` operations, there's a built-in `strip` function, use that and delete the comments

File: src/search_count_report.py [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+        return datetime.strptime(time_str, "%B %d %Y, %H:%M:%S")
```
Comment (young-k) [inline src/search_count_report.py]: can we add a comment showing what this format string looks like, or define it in a constant up top?

File: src/search_count_report.py [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+
+    def parse_day(self, timestamp):
```
Comment (young-k) [inline src/search_count_report.py]: we only call this function once and it's only one line, can we delete it

File: src/utils.py [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+    if start_index == 1:
+        return ''
+    return msg[start_index:]
+
+def parse_key(msg_json, key):
+    """This parses the value corresponding to [key] from the msg_json."""
+    key_index = msg_json.find(key)
+    if key_index == -1:
+        return ''
+    # Need to add 3 to reach the beginning of the value correspoding to this key
```
Comment (young-k) [inline src/utils.py]: `corresponding`

File: src/search_count_report.py (hunk 19 above)
Comment (kevinchan159) [reply src/search_count_report.py:44]: they're not actually strip functions

File: src/search_count_report.py [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+class SearchCountReport(Report):
+
+    def generate(self):
+        # Write fields to search_count.csv
+        self.writer.writerow(['day', 'hour', 'count'])
+
+        time_to_count_dict = {}
+
+        # Read routes.csv
+        for index, row in enumerate(self.reader):
```
Comment (young-k) [inline src/search_count_report.py:16]: you can actually do `for index, [timestamp, _] in enumerate...`, if the only variable you're accessing is `timestamp`

**Review by young-k — APPROVED** [review]
LGTM, great work with this –– left one last comment!
